x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/CombineProjections.java

```diff
@@ -137,23 +141,33 @@ private static List<NamedExpression> combineProjections(List<? extends NamedExpr
     }
 
     private static List<Expression> combineUpperGroupingsAndLowerProjections(
-        List<? extends Attribute> upperGroupings,
+        List<? extends NamedExpression> upperGroupings,
         List<? extends NamedExpression> lowerProjections
     ) {
         // Collect the alias map for resolving the source (f1 = 1, f2 = f1, etc..)
-        AttributeMap<Attribute> aliases = new AttributeMap<>();
+        AttributeMap<Expression> aliases = new AttributeMap<>();
         for (NamedExpression ne : lowerProjections) {
-            // Projections are just aliases for attributes, so casting is safe.
-            aliases.put(ne.toAttribute(), (Attribute) Alias.unwrap(ne));
+            // record the alias
+            aliases.put(ne.toAttribute(), Alias.unwrap(ne));
         }
-
         // Replace any matching attribute directly with the aliased attribute from the projection.
-        AttributeSet replaced = new AttributeSet();
-        for (Attribute attr : upperGroupings) {
-            // All substitutions happen before; groupings must be attributes at this point.
-            replaced.add(aliases.resolve(attr, attr));
+        AttributeSet seen = new AttributeSet();
+        List<Expression> replaced = new ArrayList<>();
+        for (NamedExpression ne : upperGroupings) {
+            // Duplicated attributes are ignored.
+            if (ne instanceof Attribute attribute) {
+                var newExpression = aliases.resolve(attribute, attribute);
+                if (newExpression instanceof Attribute newAttribute && seen.add(newAttribute) == false) {
+                    // Already seen, skip
+                    continue;
+                }
+                replaced.add(newExpression);
+            } else {
+                // For grouping functions, this will replace nested properties too
+                replaced.add(ne.transformUp(Attribute.class, a -> aliases.resolve(a, a)));
+            }
         }
-        return new ArrayList<>(replaced);
+        return replaced;
```

Comment:
nit: I think this change can become a bit smaller with less special-casing - which I find to be easier to reason about and verify correctness.

```
    private static List<Expression> combineUpperGroupingsAndLowerProjections(
        List<? extends NamedExpression> upperGroupings,
        List<? extends NamedExpression> lowerProjections
    ) {
        // Collect the alias map for resolving the source (f1 = 1, f2 = f1, etc..)
        AttributeMap<Attribute> aliases = new AttributeMap<>();
        for (NamedExpression ne : lowerProjections) {
            // record the alias
            aliases.put(ne.toAttribute(), (Attribute) Alias.unwrap(ne));
        }

        // Replace any matching attribute directly with the aliased attribute from the projection.
        AttributeMap<NamedExpression> replaced = new AttributeMap<>();
        for (NamedExpression ne : upperGroupings) {
            // All substitutions happen before; groupings must be attributes at this point
            // except for CATEGORIZE which will be an alias like `c = CATEGORIZE(attribute)`
            NamedExpression transformed = (NamedExpression) ne.transformUp(Attribute.class, a -> aliases.resolve(a, a));
            replaced.add(transformed.toAttribute(), transformed);
        }
        return new ArrayList<>(replaced.values());
    }
```

I ran the tests with my version and they seemed to pass the same :)